src/reducers/existingThoughtChange.js

```diff
@@ -51,12 +52,14 @@ export default (state, { oldValue, newValue, context, showContexts, thoughtsRank
     ? contextOf(contextOf(thoughtsRanked)).concat({ value: oldValue, rank: headRank(contextOf(thoughtsRanked)) }).concat(head(thoughtsRanked))
     : contextOf(thoughtsRanked).concat({ value: oldValue, rank })
 
+  const oldPath = rankThoughtsFirstMatch(state, thoughtsOld)
```

Comment:
`rankThoughtsFirstMatch` is off limits. It merely guesses at the ranks, and should only be used for restoring the cursor from a lossy url.

Reply:
Oh! I understand now. 

> @shresthabijay Can you clarify: Where are you storing the id? I assume you want to store it in one place and then look it up as needed.
> 
> * thoughtIndex
> * contextIndex
> * cursor

I have stored it in `thoughtIndex` and `contextIndex` and I persist them only when thought is moved or changed.

Reply:
> `rankThoughtsFirstMatch` is off limits. It merely guesses at the ranks, and should only be used for restoring the cursor from a lossy url.

Oh! I thought it was used to get the exact thought from the given context. So should I use `getThoughts` and then use `find` to match the exact thought ?

Reply:
Hmmm I see now that `rankThoughtsFirstMatch` already snuck into this function. This isn't really what we want here, but since it's already in `dev` we can refactor it in a separate PR.

https://github.com/cybersemics/em/commit/1c74cc5f5097e42a7ca35901c10fe92aab0677b2#diff-f3709f42ab08e75fdd90c210ec93a4e3R62-R63

Reply:
Can you get id from `thoughtOld`? You shouldn't have to do any iteration.

Reply:
Sure I will do that!

Reply:
@raineorshine Since id is associated with individual thought we need to iterate over `thoughtOld`'s `contexts` to find the exact thought.

Reply:
That's right. Okay. At least see if there is any way to utilize existing computations in the function to avoid an extra iteration.

Reply:
@raineorshine Sure! Is there other way we can get exact ranked path instead of using `rankedThoughtsFirstMatch` ? We can simply get the head of the ranked path to access the exact thought. We won't have to do extra iteration over.

Reply:
The only good way would be to actually store the `id` in the cursor. I was hoping to avoid that since there is so much logic that works with the cursor. Preserving the id there is non-trivial.

`existingThoughtChange` etc intentionally takes a `Child` and a `Context` rather than a `Path`. i.e. it shouldn't matter how the user got there, just which thought to change.

Reply:
Yes I agree about `existingThoughtChange` taking only `Child` and `Context`. But with addition of recently edited thought we still need a `thoughtsRanked`  inside `existingThoughtChange`. Also we need to get `id` of `head` to persist.

In current code base we have `thoughtsRanked` being passed to `existingThoughtChange`. `thoughtsRanked` has `RankedThought` with `rank`, `value` and `lastUpdated`. Now it will have `id` too.  This `thoughtsRanked` is used to get `headRank` and `headValue` easily inside `existingThoughtChange`. We could use same `rankedThoughts` to get `headId` or use it for recently edited thoughts. But there is a problem with this. Normal editing from `Editable` passes `thoughtsRanked` with all the available properties `rank`, `value`, `lastUpdated` and `id`. But functions like `splitFunctions` that also calls `existingThoughtChange`, pass `thoughtsRanked` with the `head` RankedThought with just `rank` and `value`. 

https://github.com/cybersemics/em/blob/cc1b235094d7685410d128058bd4793b9b5c017c/src/reducers/splitThought.js#L50

So we don't have access to `lastUpdated` and `id` of `head` thought. This inconsistency is forcing us to find ranked path again inside `existingThoughtChange` even though we already have `rankedThoughts` passed as function paramater.

But the question is should properties like `lastUpdated` and `id` be part of `RankedThought` inside `thoughtsRanked` ? If yes then we should pass it everywhere without exception otherwise it will cause confusion. If it's a `thoughtsRanked` then should we expect it to have only `rank` and `value` or should it have all properties associated with the thought ?

Reply:
Thank you, that makes sense.

I guess when it comes down to it, I am willing to add `id` everywhere it's justified. However I want to keep the interfaces clean and there may be places where it's better to derive `id` then force a caller to know about it. It's a bit fuzzy where to draw the line. I agree we want to reduce confusion, and it should be made very clear what is required.

- In one case, `{ value, rank }` serves as the minimal information to uniquely identify a thought within a context.
- In the other case, `{ value, rank, id, created, lastUpdated }` represents the full `Child` with all of its metadata.

So:

- If we *have* the metadata (id, etc), we should be passing it through. 
- If the metadata is required by the callee, we should be passing it through, even if we need to get or calculate it.
- If the metadata is not required by the callee, and it would take additional work to get the metadata, perhaps it represents a minimal interface rather than a full `Child` node.

Thanks for the discussion. At this point, after taking the above into account, I am open to your recommendation.

Reply:
Yes. It's very fuzzy where to draw the line. Let's list out what changes we need to make when we want caller (ex: `importText.js`) to pass `thoughtsRanked` with metadata that is required by callee that needs to persists id (ex: `exisitingThoughtChange`).

- At most places caller has easy access to metadata and can pass `id`, `lastUpdated` along with `rank` and `value` to `thoughtsRanked`
- We need to make changes to some utility functions that is used to calculate `thoughtsRanked` to pass necessary metadata like `id`. ex: `addContext.js`.
- Since many places uses `cursor` to derive `thoughtsRanked` to pass to callee , so we need to make sure `cursor` is always populated with `metadata`. That means we need to make necessary changes where `setCursor` is called.
- Sometimes `setCursor` is called right after `newThoughtSubmit`. We need to find the newly added thought and then pass required metadata to make sure cursor is consistently populated with `id`.

Now on the other hand if we want to callee to handle finding metadata of thought instead of making caller to pass metadata on `thoughtsRanked` itself.

- I am assuming that except `head` of the `thoughtsRanked`  passed to `exisitingThoughtChange` all other thoughts have the meta data associated with them. In this case we need to make extra iteration of `thoughtOld`'s `context`  property to find the exact thought and it's metadata. We can use it to persist `id` and create a path that has all metadata of its thought to pass it to recently edited thought.

Reply:
> * If the metadata is not required by the callee, and it would take additional work to get the metadata, perhaps it represents a minimal interface rather than a full `Child` node.

Calee like `exisitingThoughtChange` and `existingThoughtMove` are responsible for persisting metadata like `id`. So every caller that calls these functions need to make sure that they pass `thoughtsRanked` with all metadata. We just need to make changes to these callers that call these specific functions.

Reply:
Thank you, this has given me the chance to think about this more, although I am led to make the counter recommendation.
> 
> * At most places caller has easy access to metadata and can pass `id`, `lastUpdated` along with `rank` and `value` to `thoughtsRanked`
> * We need to make changes to some utility functions that is used to calculate `thoughtsRanked` to pass necessary metadata like `id`. ex: `addContext.js`.
> * Since many places uses `cursor` to derive `thoughtsRanked` to pass to callee , so we need to make sure `cursor` is always populated with `metadata`. That means we need to make necessary changes where `setCursor` is called.
> * Sometimes `setCursor` is called right after `newThoughtSubmit`. We need to find the newly added thought and then pass required metadata to make sure cursor is consistently populated with `id`.

These changes are extensive and add tight coupling to the id across multiple modules. I still think we should avoid this. The more minimal solution of deriving the id as needed gives us a minimal implementation and defers the additional risk. I would be more comfortable considering the more extensive implementation in a refactor down the road after we have been using ids with the new render pipeline for some time.

Another possibility is that we *only* pass id, and not pass value or rank at all, looking those up as needed. That's a huge change, but it should be considered. Deferring the decisions gives us more time to evaluate the merits of that approach without holding up the flat render task.

> Now on the other hand if we want callee to handle finding metadata of thought instead of making caller to pass metadata on `thoughtsRanked` itself.
> 
> * I am assuming that except `head` of the `thoughtsRanked`  passed to `exisitingThoughtChange` all other thoughts have the meta data associated with them. In this case we need to make extra iteration of `thoughtOld`'s `context`  property to find the exact thought and it's metadata. We can use it to persist `id` and create a path that has all metadata of its thought to pass it to recently edited thought.

Yes, and it will get better since I expect a refactor where `thoughtObject.context` becomes an object rather than an array so we can do O(1) lookup by value + rank.

Does recently edited need the id? I thought it was the flat render that needed it.

> * We can use a function that gives proper ranked path when given a context. But it will just be too much to do that and it doesn't make sense when we already have `thoughtsRanked` as parameter.

A Path cannot, even in theory, be deterministically derived from a Context, since there are many Paths that may lead to a Context.

> The later solution is easy to implement and requires less changes but it still doesn't solve inconsistency on the `thoughtsRanked` we are receiving from the caller. It will create confusion on why are we finding the metadata specifically for the `head` thought. So I would suggest we take the first approach. Let me know what you think.

I would rather a little confusion in one or two places than an extensive interface change across the app that we, or at least I, am yet convinced is the optimal architecture.

Reply:
> Another possibility is that we _only_ pass id, and not pass value or rank at all, looking those up as needed. That's a huge change, but it should be considered. Deferring the decisions gives us more time to evaluate the merits of that approach without holding up the flat render task.

Yes I agree with you. Since id can uniquely identify thoughts, so it makes sense to go for this approach.

Reply:
> A Path cannot, even in theory, be deterministically derived from a Context, since there are many Paths that may lead to a Context.

Ahh yes. So the problem still remains that we don't have all metadata inside `thoughtsRanked` passed inside `existingThoughtChange` and we cannot get full path just from `context` to derive exact id of head thought. So calculating full path inside `exisitingThoughtChange` can be ruled out. 

Either we pass `id` from all the callers or we go with the plan you suggested to pass only id and instead use it to derive exact thought.

Reply:
> So calculating full path inside exisitingThoughtChange can be ruled out.

But `existingThoughtChange` is already passed the full path: `thoughtsRanked`. It may have been shoehorned in, but it's there. I suggest we use it for now. Unless this isn't what you're needing?

> Either we pass id from all the callers or we go with the plan you suggested to pass only id and instead use it to derive exact thought.

Interesting, those were the two options I ruled out 😂. I was suggesting that the least risky approach right now would be to pass the id only where needed (not all callers) or derive it when possible.

Reply:
@raineorshine Let's start from where we left last time. I am still not sure about which implementation we are going for :joy: . I think for now we should just derive `id` from `contextIndex` inside `exisitingThoughtsChange` and `exisitingThoughtsMove`. It would only require us making changes to these files.

Reply:
Yes, let's go with that for now! Apologies for the confusion.